btree/btree.c

```diff
@@ -3865,19 +3879,27 @@ static int64_t btree_put_kv_tick(struct m0_sm_op *smop)
 			}
 		}
 		/** Fall through if path_check is successful. */
-	case P_MAKESPACE: {
-		if (oi->i_key_found) {
-			struct m0_btree_rec rec;
-			rec.r_flags = M0_BSC_KEY_EXISTS;
-			int rc = bop->bo_cb.c_act(&bop->bo_cb, &rec);
-			if (rc) {
-				lock_op_unlock(tree);
-				return fail(bop, rc);
-			}
+	case P_SANITY_CHECK: {
+		int  rc;
+		if (bop->bo_opc == M0_BO_PUT && oi->i_key_found)
+			rc = M0_BSC_KEY_EXISTS;
+		else if (bop->bo_opc == M0_BO_UPDATE && !oi->i_key_found)
+			rc = M0_BSC_KEY_NOT_FOUND;
+		else
+			rc = M0_BSC_SUCCESS;
+
+		if (rc != M0_BSC_SUCCESS) {
+			oi->i_sm_rc = rc;
```

Comment:
I do not see the use of `i_sm_rc` variable. It is getting updated when `rc != M0_BSC_SUCCESS` and from here tick sm will exit via fail routine.

Reply:
Yes in P_FINI, we are checking if sm_rc is non-zero, if it is non-zero, we are returning ERROR

Reply:
In fail routine `bop->bo_op.o_sm.sm_rc = rc` will update the state machine return code, we don't need the error returning code in P_FINI.

Reply:
bop->bo_op.o_sm.sm_rc  will get reinitialized to 0  in op.c: m0_sm_op_tick() after returning next state . That's why we can not use bop->bo_op.o_sm.sm_rc. I also thought the same but because of this issue, had to add one more flag in oimpl structure.